doc/mgmt/SONiC_YANG_Model_Guidelines.md

```diff
@@ -0,0 +1,961 @@
+
+#### YANG
+```
+	leaf SRC_IP {
+		type inet:ipv4-prefix; <<<<
+	}
+
+	leaf DST_IP {
+		type inet:ipv4-prefix;
+	}
+```
+
+### 9. Data Node/Object Hierarchy of the an objects in YANG models will be same as for all the fields at same hierarchy in Config DB. If any exception is created then it must be recorded properly with comment under object level in YANG models. To see an example of a comment, please refer to the step with heading as "Comment all must, when and patterns conditions." below.
+
+For Example:
+
+"Family" of VLAN_INTERFACE and "IP_TYPE" of ACL_RULE should be at same level in YANG model too.
+
+#### ABNF
+```
+"VLAN_INTERFACE": {
+        "Vlan100|2a04:f547:45:6709::1/64": {
+            "scope": "global",
+            "family": "IPv6"
+        }
+}
+"ACL_RULE": {
+        "NO-NSW-PACL-V4|DEFAULT_DENY": {
+            "PACKET_ACTION": "DROP",
+            "IP_TYPE": "IPv4ANY",
+            "PRIORITY": "0"
+        }
+}
+```
+####  YANG
+In YANG, "Family" of VLAN_INTERFACE and "IP_TYPE" of ACL_RULE is at same level.
+```
+container VLAN_INTERFACE {
+        description "VLAN_INTERFACE part of config_db.json";
+        list VLAN_INTERFACE_LIST {
+            ......
+            ......
+            leaf family {
+                type sonic-head:ip-family;
+            }
+        }
+}
+
+container ACL_RULE {
+        description "ACL_RULE part of config_db.json";
+        list ACL_RULE_LIST {
+            ......
+            ......
+            leaf IP_TYPE {
+                type sonic-head:ip_type;
+            }
+    }
+}
+```
+
+### 10.  If an object is part of primary-key in ABNF.json, then it should be a key in YANG model. In YANG models, a primary-key from ABNF.json can be represented either as name of a Container object or as a key field in List object. Exception must be recorded in YANG Model with a comment in object field. To see an example of a comment, please refer to the step with heading as "Comment all must, when and patterns conditions." below. Though, key names are not stored in Redis DB, use the same key name as defined in ABNF schema.
+
+Example: VLAN_MEMBER dictionary in ABNF.json has both vlan-id and ifname part of the key. So YANG model should have the same keys.
+
+#### ABNF
+```
+ "VLAN_MEMBER": {
+        "Vlan100|Ethernet0": { //<<<< KEYS
+            "tagging_mode": "untagged"
+        }
+   }
+```
+;Defines interfaces which are members of a vlan
+key                 = VLAN_MEMBER_TABLE:"Vlan"vlanid:ifname ;
+
+#### YANG
+```
+container VLAN_MEMBER {
+    description "VLAN_MEMBER part of config_db.json";
+        list ..... {
+            key "vlan-name ifname";//<<<< KEYS
+       }
+}
+```
+
+### 11. If any key used in current table refers to other table, use leafref type for the key leaf definition. 
+
+Example : 
+#### ABNF
+
+```
+key: ACL_RULE_TABLE:table_name:rule_name
+.....
+.....
+```
+#### YANG
+```
+.....
+container ACL_TABLE {
+	list ACL_TABLE_LIST {
+		key aclname;
+			
+		leaf aclname {
```

Comment:
I guess, table_name instead of aclname, if we keep Same key name as ABNF.

Reply:
Will use "name" as per the ABNF schema in https://github.com/Azure/SONiC/wiki/ACL-High-Level-Design.

Reply:
Done.